### ISSUE TITLE
enhancement: kpatch.spec should install systemd or upstart files

### ISSUE THREAD

**Issue by joe-lawrence** [dynup/kpatch#891]
The current [contrib/kpatch.spec](https://github.com/dynup/kpatch/blob/master/contrib/kpatch.spec) file will install both systemd and upstart configuration files -- this may be confusing to end users.  It would be great if the [Makefile](https://github.com/dynup/kpatch/blob/master/contrib/Makefile) / specfile were smart enough to install only the necessary file.

Googling "detect systemd or upstart" eventually brought me to the manpage for [sd_booted()](http://manpages.ubuntu.com/manpages/xenial/en/man3/sd_booted.3.html).  It says,

> Internally, this function checks whether the directory /run/systemd/system/ exists. A simple check like this can also be implemented trivially in shell or any other language.

I can't think of any elegant way to handle conditional specfile `%files` entries aside from pre-processing the specfile.  But in RHEL, we do not directly use the [contrib/kpatch.spec](https://github.com/dynup/kpatch/blob/master/contrib/kpatch.spec) file anyway.  It would be interesting to know how other distributions are using it, and if it is only serving as an example.  If so, then perhaps it would be simpler to modify the [Makefile](https://github.com/dynup/kpatch/blob/master/contrib/Makefile) with `systemd` and `upstart` targets, then everyone's specfile can choose the desired target(s) and specify the proper `%files` section.

/cc @aliceinwire (gentoo) @arges (ubuntu) @kirawrath (AWS?)

**Comment by kirawrath** [comment]
Hello there!

I am afraid this may be trickier than it seems in a first glance. This behavior was introduced in [this commit](https://github.com/dynup/kpatch/pull/812), and in the thread you can see that at first I tried not to install it directly in `/etc/init`, however jpoimboe rightly asked me to install it unconditionally, which I completely agree.

The main issues in installing it conditionally are:
- **There is no standard fail-proof way to detect if a system is using upstart or not.** The best you can do is to try some 3 different methods which would cover 99% of the cases, but will inevitably fail in a more esoteric system.
- **Some systems have both upstart and systemd installed.** I have seen myself a few times grub entries what would boot the OS either in systemd or upstart.
- A minor issue is also the fact that I would rather keep the installation process as simple as possible, and to me installing it conditionally seem to add unnecessary complexity (however small).

Now, when we look at the disadvantages or installing the config file unconditionally is just that it will use a few extra bytes of HDD, however systems that employ `kpatch` probably would not mind those bytes.

You mentioned that this may be confusing to users, but what kind of users are we talking here? I am sure the majority would not bother in looking up for those files.

Now, this is just how I see this issue, which may or may not be the best approach, and I would gladly change my mind in light of more arguments.

**Comment by arges** [comment]
Looking at how debian packaging does this and specifically dh_installinit
which in turns invokes update-rc.d:
https://git.launchpad.net/ubuntu/+source/init-system-helpers/tree/script/update-rc.d?h=debian/sid#n91
So seems like checking for /run/systemd/system isn't a bad way to check.

However, I've noticed that some debian packages will install both upstart,
systemd, (and even sysvinit scripts).
The init-system-helpers package in debian in turn provides scripts that
determine which files to use to run the service (if you use something like
invoke-rc.d/service).
So if you type 'service daemon start', and have systemd and upstart files
installed, and you are running systemd, the systemd unit is used.

tldr; installing multiple service files in debian/ubuntu is not an issue.

--chris

**Comment by joe-lawrence** [comment]
Sorry for putting this on the back burner, PTO and other priorities kept me away from kpatch duties...

@kirawrath : your reasoning is sound, especially in light of the difficulties in reliably determining init systems.  I had only encountered an issue when rebasing our official Red Hat kpatch RPM, which contains its own specfile copy ... recent kpatch releases now contain an updated install target and rpmbuild complained when I hadn't updated our copy for those init scripts.  From an officially supported Red Hat kpatch RPM point of view, I didn't think it made sense to install scripts for an init system that wasn't even supported by the vendor (the customer confusion part).  One thought I had was to elaborate on the "install" target ("install_upstart", "install_systemd", "install" = install both).   Perhaps it's not worth the complications, but I figured I would at least float the idea.   This can always be solved if needed on the distribution side by modifying our internal specfile.

Also, is anyone actually using the contrib/kpatch.spec file (directly or indirectly)?

Thanks @arges for debian explanation!

**Comment by joe-lawrence** [comment]
Closing due to inactivity.
